code/go/internal/validator/folder_spec.go

```diff
@@ -71,44 +71,51 @@ func (s *folderSpec) load(fs fs.FS, specPath string) error {
 	return nil
 }
 
-func (s *folderSpec) validate(packageName string, fsys fspath.FS, path string) ve.ValidationErrors {
+func (s *folderSpec) validate(pkg *Package, path string) ve.ValidationErrors {
 	var errs ve.ValidationErrors
-	files, err := fs.ReadDir(fsys, path)
+	files, err := fs.ReadDir(pkg, path)
 	if err != nil {
-		errs = append(errs, errors.Wrapf(err, "could not read folder [%s]", fsys.Path(path)))
+		errs = append(errs, errors.Wrapf(err, "could not read folder [%s]", pkg.Path(path)))
 		return errs
 	}
 
 	// This is not taking into account if the folder is for development. Enforce
 	// this limit in all cases to avoid having to read too many files.
 	if contentsLimit := s.Limits.TotalContentsLimit; contentsLimit > 0 && len(files) > contentsLimit {
-		errs = append(errs, errors.Errorf("folder [%s] exceeds the limit of %d files", fsys.Path(path), contentsLimit))
+		errs = append(errs, errors.Errorf("folder [%s] exceeds the limit of %d files", pkg.Path(path), contentsLimit))
 		return errs
 	}
 
+	// Don't enable beta features for packages marked as GA.
+	if s.Release == "beta" && pkg.Version.Prerelease() == "" {
+		errs = append(errs, errors.Errorf("spec for [%s] defines beta features which can't be enabled for packages with a stable semantic version", pkg.Path(path)))
+	} else if s.Release == "beta" && pkg.Version.Prerelease() != "" {
+		log.Printf("Warning: package with non-stable semantic version and active beta features (enabled in [%s]) can't be released as stable version.q", pkg.Path(path))
```

Comment:
We need some way to return different "levels" of errors. A developer may want to have strict validation that fails on warnings too.

Reply:
Issue created: https://github.com/elastic/package-spec/issues/342